File: internal/expr.go
package internal

type expr interface {
	accept(exprVisitor) R
}

type exprVisitor interface {
	visitListExpr(expr *listExpr) R
	visitDictionaryExpr(expr *dictionaryExpr) R
	visitAssignExpr(expr *assignExpr) R
	visitAccessExpr(expr *accessExpr) R
	visitBinaryExpr(expr *binaryExpr) R
	visitCallExpr(expr *callExpr) R
	visitGetExpr(expr *getExpr) R
	visitSetExpr(expr *setExpr) R
	visitSuperExpr(expr *superExpr) R
	visitGroupingExpr(expr *groupingExpr) R
	visitLiteralExpr(expr *literalExpr) R
	visitLogicalExpr(expr *logicalExpr) R
	visitThisExpr(expr *thisExpr) R
	visitUnaryExpr(expr *unaryExpr) R
	visitVariableExpr(expr *variableExpr) R
	visitFunctionExpr(expr *functionExpr) R
}

type listExpr struct {
	elements []expr
	brace *token
}

func (s *listExpr) accept(visitor exprVisitor) R {
	return visitor.visitListExpr(s)
}

type dictionaryExpr struct {
	elements []expr
	curlyBrace *token
}

func (s *dictionaryExpr) accept(visitor exprVisitor) R {
	return visitor.visitDictionaryExpr(s)
}

type assignExpr struct {
	name *token
	value expr
}

func (s *assignExpr) accept(visitor exprVisitor) R {
	return visitor.visitAssignExpr(s)
}

type accessExpr struct {
	object expr
	brace *token
	first expr
	firstColon *token
	second expr
	secondColon *token
	third expr
}

func (s *accessExpr) accept(visitor exprVisitor) R {
	return visitor.visitAccessExpr(s)
}

type binaryExpr struct {
	left expr
	operator *token
	right expr
}

func (s *binaryExpr) accept(visitor exprVisitor) R {
	return visitor.visitBinaryExpr(s)
}

type callExpr struct {
	callee expr
	paren *token
	arguments []expr
}

func (s *callExpr) accept(visitor exprVisitor) R {
	return visitor.visitCallExpr(s)
}

type getExpr struct {
	object expr
	name *token
}

func (s *getExpr) accept(visitor exprVisitor) R {
	return visitor.visitGetExpr(s)
}

type setExpr struct {
	object expr
	name *token
	value expr
}

func (s *setExpr) accept(visitor exprVisitor) R {
	return visitor.visitSetExpr(s)
}

type superExpr struct {
	keyword *token
	method *token
}

func (s *superExpr) accept(visitor exprVisitor) R {
	return visitor.visitSuperExpr(s)
}

type groupingExpr struct {
	expression expr
}

func (s *groupingExpr) accept(visitor exprVisitor) R {
	return visitor.visitGroupingExpr(s)
}

type literalExpr struct {
	value interface{}
}

func (s *literalExpr) accept(visitor exprVisitor) R {
	return visitor.visitLiteralExpr(s)
}

type logicalExpr struct {
	left expr
	operator *token
	right expr
}

func (s *logicalExpr) accept(visitor exprVisitor) R {
	return visitor.visitLogicalExpr(s)
}

type thisExpr struct {
	keyword *token
}

func (s *thisExpr) accept(visitor exprVisitor) R {
	return visitor.visitThisExpr(s)
}

type unaryExpr struct {
	operator *token
	right expr
}

func (s *unaryExpr) accept(visitor exprVisitor) R {
	return visitor.visitUnaryExpr(s)
}

type variableExpr struct {
	name *token
}

func (s *variableExpr) accept(visitor exprVisitor) R {
	return visitor.visitVariableExpr(s)
}

type functionExpr struct {
	params []*token
	body []stmt
}

func (s *functionExpr) accept(visitor exprVisitor) R {
	return visitor.visitFunctionExpr(s)
}

File: rewrite_in_rust/src/expr.rs
use crate::stmt::*;
use crate::token::*;

#[derive(Debug, PartialEq, Clone)]
pub struct FnExpr {
    pub params: Vec<TokenData>,
    pub body: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct VarExpr {
    pub name: Option<TokenData>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ListExpr {
    pub elements: Vec<Expr>,
    pub brace: TokenData,
}

#[derive(Debug, PartialEq, Clone)]
pub struct DictionaryExpr {
    pub elements: Vec<Expr>,
    pub curly_brace: TokenData,
}

#[derive(Debug, PartialEq, Clone)]
pub struct AssignExpr {
    pub name: TokenData,
    pub value: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct AccessExpr {
    pub object: Box<Expr>,
    pub brace: TokenData,
    pub first: Box<Expr>,
    pub first_colon: TokenData,
    pub second: Box<Expr>,
    pub second_colon: TokenData,
    pub third: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct BinaryExpr {
    pub left: Box<Expr>,
    pub operator: TokenData,
    pub right: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct CallExpr {
    pub callee: Box<Expr>,
    pub paren: TokenData,
    pub arguments: Vec<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct GetExpr {
    pub object: Box<Expr>,
    pub name: TokenData,
}

#[derive(Debug, PartialEq, Clone)]
pub struct SetExpr {
    pub object: Box<Expr>,
    pub name: TokenData,
    pub value: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct SuperExpr {
    pub keyword: TokenData,
    pub method: TokenData,
}

#[derive(Debug, PartialEq, Clone)]
pub struct GroupingExpr {
    pub expression: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct LiteralExpr {
    pub value: Literal,
}

#[derive(Debug, PartialEq, Clone)]
pub struct LogicalExpr {
    pub left: Box<Expr>,
    pub operator: TokenData,
    pub right: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ThisExpr {
    pub keyword: TokenData,
}

#[derive(Debug, PartialEq, Clone)]
pub struct UnaryExpr {
    pub operator: TokenData,
    pub right: Box<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Expr {
    Fn(FnExpr),
    Var(VarExpr),
    List(ListExpr),
    Dictionary(DictionaryExpr),
    Assign(AssignExpr),
    Access(AccessExpr),
    Binary(BinaryExpr),
    Call(CallExpr),
    Get(GetExpr),
    Set(SetExpr),
    Super(SuperExpr),
    Grouping(GroupingExpr),
    Literal(LiteralExpr),
    Logical(LogicalExpr),
    This(ThisExpr),
    Unary(UnaryExpr),
    Empty,
}

impl Expr {
    pub fn is_empty(&self) -> bool {
        match self {
            Expr::Empty => true,
            _ => false,
        }
    }
}

pub trait ExprVisitor<T> {
    fn visit_function_expr(&mut self, expr: &FnExpr) -> T;
    fn visit_variable_expr(&mut self, expr: &VarExpr) -> T;
    fn visit_list_expr(&mut self, expr: &ListExpr) -> T;
    fn visit_dictionary_expr(&mut self, expr: &DictionaryExpr) -> T;
    fn visit_assign_expr(&mut self, expr: &AssignExpr) -> T;
    fn visit_access_expr(&mut self, expr: &AccessExpr) -> T;
    fn visit_binary_expr(&mut self, expr: &BinaryExpr) -> T;
    fn visit_call_expr(&mut self, expr: &CallExpr) -> T;
    fn visit_get_expr(&mut self, expr: &GetExpr) -> T;
    fn visit_set_expr(&mut self, expr: &SetExpr) -> T;
    fn visit_super_expr(&mut self, expr: &SuperExpr) -> T;
    fn visit_grouping_expr(&mut self, expr: &GroupingExpr) -> T;
    fn visit_literal_expr(&mut self, expr: &LiteralExpr) -> T;
    fn visit_logical_expr(&mut self, expr: &LogicalExpr) -> T;
    fn visit_this_expr(&mut self, expr: &ThisExpr) -> T;
    fn visit_unary_expr(&mut self, expr: &UnaryExpr) -> T;
}

pub trait ExprAcceptor<T> {
    fn accept(&self, visitor: &mut dyn ExprVisitor<T>) -> T;
}
